### PR TITLE
Document and `scratch` directory for sqllogictest and make test specific

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?
Follow on to https://github.com/apache/arrow-datafusion/pull/7283

## Rationale for this change

@devinjdangelo added a neat `scratch` concept to sqllogictest but it is not yet documented. See discussion at https://github.com/apache/arrow-datafusion/pull/7283#discussion_r1294546967

## What changes are included in this PR?
1. Change scratch dir to be test name specific (as the tests are run in parallel)
2. Update `copy.slt` to use the specific direcotry
1. Document `scratch` dir in README
 
## Are these changes tested?
Yes

### CODE REVIEW

File: datafusion/sqllogictest/README.md
Comment (devinjdangelo) [inline datafusion/sqllogictest/README.md:186]: This looks great to me! I ran into some issues needing to make sure each table had a unique subpath betwen copy/insert tests in #7294, and this would simplify that.

Comment (alamb) [reply datafusion/sqllogictest/README.md:186]: Merged https://github.com/apache/arrow-datafusion/pull/7294 and I updated this PR with the changes from there

**Comment by alamb** [comment]
@Dandandan / @jackwener  / @liukun4515  here is another example where the PR is ready to merge but I can't merge it due to the check added in https://github.com/apache/arrow-datafusion/pull/7226

Can another committer please approve this PR?

**Review by yjshen — APPROVED** [review]

**Comment by alamb** [comment]
Thanks @yjshen ❤️

**Pull request merged**
